File: Plutus/Frontend/Plutus.Frontend.WebApp/src/openingHours.ts
/**
 * The store's opening hours, as the portal stores them.
 *
 * ⚠⚠ WHY THIS EXISTS. Both tills rendered *"Not set — add opening hours in the management portal"*
 * for **three different situations**: genuinely not set, present but malformed, and present in a
 * shape the reader did not expect. Matt, 2026-08-17: *"webtill does not show the opening hours set in
 * the portal"* — and from that message nobody could tell which of the three it was, because the
 * screen said the same thing for all of them. **A screen that reports one sentence for three states
 * is not information, it is a shrug.**
 *
 * ⚠⚠ AND THE PORTAL HAS AN UNVALIDATED "Advanced (JSON)" TEXTAREA. `StoresPage.tsx`'s hours editor
 * offers a raw textarea whose `onChange` sends the typed string straight to the server — a trailing
 * comma, single quotes or `{mon: […]}` all save happily. The portal then shows the operator their own
 * text back, so the portal looks *set* while every till reads it as *not set*. That is precisely the
 * shape of what was reported, and it is why this parser is TOLERANT and, when it cannot cope, SAYS SO.
 *
 * ⚠ C2 TWIN of `Plutus.Client.Core/OpeningHours.cs`. Same tolerances, same three states, same
 * wording — a shop whose hours read fine on one till and "not set" on the other is the same class of
 * defect, just cheaper than money. `till-design.md` C2 pins the pair.
 *
 * ⚠ NOT money, so it fails OPEN: anything it cannot understand is reported to the operator as
 * unreadable and the rest of the screen renders normally. A store-details screen must never be the
 * thing that takes a till down.
 */

export interface OpeningSpan {
  open: string;
  close: string;
}

export interface OpeningDay {
  /** The portal's own key — `mon`…`sun`. */
  key: string;
  label: string;
  /** Empty means CLOSED that day — not "unknown". */
  spans: OpeningSpan[];
}

export type OpeningHours =
  /** The portal has hours for this store; render the week. */
  | { state: "set"; week: OpeningDay[] }
  /** Nothing stored. The honest answer, and the only one that should send somebody to the portal. */
  | { state: "unset" }
  /** Something IS stored and this cannot read it. ⚠ Never silently shown as "unset". */
  | { state: "unreadable"; detail: string };

/** ⚠ The portal's keys in the portal's own order — never `Date`'s day numbering, which starts on
 *  Sunday and would silently reorder a shop's week. */
export const DAYS: readonly (readonly [string, string])[] = [
  ["mon", "Monday"], ["tue", "Tuesday"], ["wed", "Wednesday"], ["thu", "Thursday"],
  ["fri", "Friday"], ["sat", "Saturday"], ["sun", "Sunday"],
];

/**
 * A day key as the portal writes it, from whatever a human typed.
 *
 * ⚠ `Monday`, `MON`, ` mon ` and `monday` all mean Monday. The simple editor writes `mon`; the
 * advanced textarea is typed by a person, and refusing their whole week over a capital letter would
 * be pedantry with a shop's opening times as the cost.
 */
function dayKey(raw: string): string | null {
  const k = raw.trim().toLowerCase().slice(0, 3);
  return DAYS.some(([key]) => key === k) ? k : null;
}

/**
 * A time, normalised for display.
 *
 * ⚠ `9:00` → `09:00` so the column lines up, and `09:00:00` → `09:00` because a shop's door does not
 * open on a second boundary. Anything else is passed through untouched rather than rejected — a
 * time this does not recognise is still better shown than swallowed.
 */
function time(raw: unknown): string | null {
  if (typeof raw !== "string") return null;
  const t = raw.trim();
  if (!t) return null;
  const m = /^(\d{1,2}):(\d{2})(?::\d{2})?$/.exec(t);
  return m ? `${m[1].padStart(2, "0")}:${m[2]}` : t;
}

/**
 * One span, from the several shapes a person writes it in.
 *
 * ⚠ `open`/`close` are the portal's names; `from`/`to` are what somebody types when they have not
 * looked. ⚠ A plain string `"09:00-17:30"` is accepted too (en dash included — a portal field
 * pasted out of a document carries one).
 */
function span(raw: unknown): OpeningSpan | null {
  if (typeof raw === "string") {
    const parts = raw.split(/[-–—to]+/i).map((p) => p.trim()).filter(Boolean);
    if (parts.length !== 2) return null;
    const open = time(parts[0]);
    const close = time(parts[1]);
    return open && close ? { open, close } : null;
  }

  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;

  // ⚠ Case-insensitive field names for the same reason the keys are: this half of the blob can be
  // hand-typed.
  const lowered: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(raw as Record<string, unknown>)) lowered[k.toLowerCase()] = v;

  const open = time(lowered.open ?? lowered.from ?? lowered.start);
  const close = time(lowered.close ?? lowered.to ?? lowered.end);
  return open && close ? { open, close } : null;
}

/** A day's value, which may be one span, several, or a string. */
function spansOf(raw: unknown): { spans: OpeningSpan[]; hadContent: boolean } {
  if (raw === null || raw === undefined) return { spans: [], hadContent: false };

  // ⚠ `false` / `"closed"` are how a person says shut. Treated as an explicit CLOSED, not as junk.
  if (raw === false) return { spans: [], hadContent: true };
  if (typeof raw === "string" && /^\s*(closed|shut|)\s*$/i.test(raw)) return { spans: [], hadContent: true };

  const list = Array.isArray(raw) ? raw : [raw];
  if (list.length === 0) return { spans: [], hadContent: true };   // `[]` IS "closed", deliberately

  const spans = list.map(span).filter((s): s is OpeningSpan => s !== null);
  return { spans, hadContent: true };
}

/**
 * Read the portal's blob.
 *
 * ⚠⚠ THE THREE STATES ARE THE POINT. `unset` sends somebody to the portal; `unreadable` tells them
 * the portal already holds something and it needs looking at. Collapsing them — which is what both
 * tills did until 2026-08-17 — means a shop with mistyped hours is told to type them again into the
 * same box that is already wrong.
 */
export function parseOpeningHours(json: string | null | undefined): OpeningHours {
  if (!json || !json.trim()) return { state: "unset" };

  let raw: unknown;
  try {
    raw = JSON.parse(json);
  } catch (e) {
    // ⚠ The parser's own message, verbatim. "Invalid JSON" tells nobody which character to look at.
    return { state: "unreadable", detail: e instanceof Error ? e.message : String(e) };
  }

  if (raw === null) return { state: "unset" };   // a stored literal `null` is nothing stored

  if (typeof raw !== "object" || Array.isArray(raw))
    return {
      state: "unreadable",
      detail: `expected an object of days, e.g. {"mon":[{"open":"09:00","close":"17:30"}]} — got ${
        Array.isArray(raw) ? "a list" : typeof raw}`,
    };

  const entries = Object.entries(raw as Record<string, unknown>);
  if (entries.length === 0) return { state: "unset" };

  const byKey = new Map<string, OpeningSpan[]>();
  const unknownKeys: string[] = [];
  let recognisedContent = false;
  let lostContent = false;

  for (const [rawKey, value] of entries) {
    const key = dayKey(rawKey);
    if (key === null) {
      unknownKeys.push(rawKey);
      continue;
    }
    const { spans, hadContent } = spansOf(value);
    // ⚠ MERGED, not replaced. `{"mon":…,"Monday":…}` is a mistake somebody could make in the
    // advanced box, and dropping the second one would lose an afternoon without saying so.
    byKey.set(key, [...(byKey.get(key) ?? []), ...spans]);
    if (hadContent) recognisedContent = true;
    if (hadContent && spans.length === 0 && value !== false && !Array.isArray(value) && typeof value !== "string")
      lostContent = true;   // an object we could not read as a span — do not pass it off as "closed"
  }

  if (unknownKeys.length > 0 && byKey.size === 0)
    return {
      state: "unreadable",
      detail: `none of these are days: ${unknownKeys.slice(0, 7).join(", ")} — expected mon, tue, wed, thu, fri, sat, sun`,
    };

  if (lostContent && [...byKey.values()].every((s) => s.length === 0))
    return { state: "unreadable", detail: 'each day needs times, e.g. {"open":"09:00","close":"17:30"}' };

  if (!recognisedContent) return { state: "unset" };

  // ⚠ A partly-understood blob still renders — five correct days plus `unknownDayKeys` naming the
  // leftovers beats showing nothing at all.
  return { state: "set", week: DAYS.map(([key, label]) => ({ key, label, spans: byKey.get(key) ?? [] })) };
}

/** What a day reads as on screen. ⚠ "Closed" — never blank, which looks like a broken binding. */
export const dayText = (d: OpeningDay): string =>
  d.spans.length === 0 ? "Closed" : d.spans.map((s) => `${s.open}–${s.close}`).join(", ");

/** ⚠ The keys the portal sent that are not days, so a screen can name them rather than drop them
 *  silently. Empty for a clean blob. */
export function unknownDayKeys(json: string | null | undefined): string[] {
  if (!json || !json.trim()) return [];
  try {
    const raw = JSON.parse(json);
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) return [];
    return Object.keys(raw as Record<string, unknown>).filter((k) => dayKey(k) === null);
  } catch {
    return [];
  }
}

File: src/Plutus.Client.Core/OpeningHours.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace Plutus.Client.Core;

/// <summary>How a store's opening hours read.</summary>
public enum OpeningHoursState
{
    /// <summary>Nothing stored. ⚠ The ONLY state that should send somebody to the portal to add them.</summary>
    Unset = 0,
    /// <summary>Hours are stored and understood.</summary>
    Set = 1,
    /// <summary>Something IS stored and this cannot read it. ⚠ Never shown as <see cref="Unset"/>.</summary>
    Unreadable = 2,
}

/// <summary>One open period. ⚠ Times are for DISPLAY — normalised to `HH:mm`, never re-parsed as money.</summary>
public sealed record OpeningSpan(string Open, string Close);

/// <summary>One day. ⚠ No spans means CLOSED, not "unknown".</summary>
public sealed record OpeningDay(string Key, string Label, IReadOnlyList<OpeningSpan> Spans);

/// <summary>The reading. <see cref="Week"/> is empty unless <see cref="State"/> is
/// <see cref="OpeningHoursState.Set"/>; <see cref="Detail"/> is filled only when it is
/// <see cref="OpeningHoursState.Unreadable"/>.</summary>
public sealed record OpeningHoursReading(
    OpeningHoursState State, IReadOnlyList<OpeningDay> Week, string? Detail);

/// <summary>
/// The store's opening hours, as the portal stores them.
///
/// ⚠⚠ WHY THIS EXISTS. Both tills printed <i>"Not set — add opening hours in the management portal"</i>
/// for <b>three different situations</b>: genuinely not set, present but malformed, and present in a
/// shape the reader did not expect. Matt, 2026-08-17: <i>"webtill does not show the opening hours set
/// in the portal"</i> — and from that message nobody could tell which of the three it was, because both
/// screens said the same thing for all of them. <b>One sentence for three states is not information,
/// it is a shrug.</b>
///
/// ⚠⚠ AND THE PORTAL HAD AN UNVALIDATED "Advanced (JSON)" TEXTAREA, whose contents went straight to
/// the server. A trailing comma or an unquoted key saved happily, the portal showed the operator their
/// own text back, and every till read it as "not set". That is the shape of what was reported, which is
/// why this parser is TOLERANT and, when it cannot cope, SAYS SO. The portal is now the strict end of
/// the agreement (it refuses to save what a till would have to stretch for); the tills are the
/// tolerant end, because they have to read what is already in the field.
///
/// ⚠ C2 TWIN of the web till's <c>src/openingHours.ts</c> — same tolerances, same three states, same
/// wording. A shop whose hours read fine on one till and "not set" on the other is the same class of
/// defect as a money divergence, just cheaper. <c>till-design.md</c> C2 pins the pair.
///
/// ⚠ NOT money, so it fails OPEN: anything it cannot understand is reported and the rest of the screen
/// renders. A store-details screen must never be the thing that takes a till down.
/// </summary>
public static class OpeningHours
{
    /// <summary>⚠ The portal's keys in the portal's own order — never <see cref="DayOfWeek"/>, which
    /// starts on Sunday and would silently reorder a shop's week.</summary>
    public static readonly (string Key, string Label)[] Days =
    {
        ("mon", "Monday"), ("tue", "Tuesday"), ("wed", "Wednesday"), ("thu", "Thursday"),
        ("fri", "Friday"), ("sat", "Saturday"), ("sun", "Sunday"),
    };

    /// <summary>
    /// Read the portal's blob.
    ///
    /// ⚠⚠ THE THREE STATES ARE THE POINT. <see cref="OpeningHoursState.Unset"/> sends somebody to the
    /// portal; <see cref="OpeningHoursState.Unreadable"/> tells them the portal already holds something
    /// that needs looking at. Collapsing them — which both tills did until 2026-08-17 — means a shop
    /// with mistyped hours is told to type them again into the box that is already wrong.
    /// </summary>
    public static OpeningHoursReading Parse(string? json)
    {
        if (string.IsNullOrWhiteSpace(json)) return Unset;

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(json);
        }
        catch (JsonException ex)
        {
            // ⚠ The parser's own message. "Invalid JSON" tells nobody which character to look at.
            return new OpeningHoursReading(OpeningHoursState.Unreadable, Array.Empty<OpeningDay>(), ex.Message);
        }

        using (doc)
        {
            var root = doc.RootElement;

            if (root.ValueKind == JsonValueKind.Null) return Unset;   // a stored literal null is nothing

            if (root.ValueKind != JsonValueKind.Object)
                return Unreadable(
                    "expected an object of days, e.g. {\"mon\":[{\"open\":\"09:00\",\"close\":\"17:30\"}]} — got "
                    + (root.ValueKind == JsonValueKind.Array ? "a list" : root.ValueKind.ToString().ToLowerInvariant()));

            var properties = root.EnumerateObject().ToList();
            if (properties.Count == 0) return Unset;

            var byKey = new Dictionary<string, List<OpeningSpan>>();
            var unknownKeys = new List<string>();
            var recognisedContent = false;
            var lostContent = false;

            foreach (var property in properties)
            {
                var key = DayKey(property.Name);
                if (key is null)
                {
                    unknownKeys.Add(property.Name);
                    continue;
                }

                var (spans, hadContent, unreadableValue) = SpansOf(property.Value);

                // ⚠ MERGED, not replaced. {"mon":…,"Monday":…} is a mistake somebody could make in the
                // advanced box, and dropping the second would lose an afternoon without saying so.
                if (!byKey.TryGetValue(key, out var existing)) byKey[key] = existing = new List<OpeningSpan>();
                existing.AddRange(spans);

                if (hadContent) recognisedContent = true;
                if (unreadableValue) lostContent = true;
            }

            if (unknownKeys.Count > 0 && byKey.Count == 0)
                return Unreadable($"none of these are days: {string.Join(", ", unknownKeys.Take(7))} — "
                    + $"expected {string.Join(", ", Days.Select(d => d.Key))}");

            if (lostContent && byKey.Values.All(s => s.Count == 0))
                return Unreadable("each day needs times, e.g. {\"open\":\"09:00\",\"close\":\"17:30\"}");

            if (!recognisedContent) return Unset;

            // ⚠ A partly-understood blob still renders — the leftovers are named by
            // `UnknownDayKeys` rather than dropped, because a key nobody reads is a setting somebody
            // believes is in effect.
            var week = Days
                .Select(d => new OpeningDay(d.Key, d.Label,
                    byKey.TryGetValue(d.Key, out var s) ? s : Array.Empty<OpeningSpan>()))
                .ToList();

            return new OpeningHoursReading(OpeningHoursState.Set, week, null);
        }
    }

    /// <summary>What a day reads as on screen. ⚠ "Closed" — never blank, which on MAUI is
    /// indistinguishable from a binding to a property that does not exist.</summary>
    public static string DayText(OpeningDay day) =>
        day.Spans.Count == 0 ? "Closed" : string.Join(", ", day.Spans.Select(s => $"{s.Open}–{s.Close}"));

    /// <summary>⚠ The keys the portal sent that are not days, so a screen can NAME them rather than
    /// drop them silently. Empty for a clean blob, and empty for junk (the unreadable message covers
    /// that case).</summary>
    public static IReadOnlyList<string> UnknownDayKeys(string? json)
    {
        if (string.IsNullOrWhiteSpace(json)) return Array.Empty<string>();
        try
        {
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return Array.Empty<string>();
            return doc.RootElement.EnumerateObject()
                .Where(p => DayKey(p.Name) is null)
                .Select(p => p.Name)
                .ToList();
        }
        catch (JsonException)
        {
            return Array.Empty<string>();
        }
    }

    private static OpeningHoursReading Unset { get; } =
        new(OpeningHoursState.Unset, Array.Empty<OpeningDay>(), null);

    private static OpeningHoursReading Unreadable(string detail) =>
        new(OpeningHoursState.Unreadable, Array.Empty<OpeningDay>(), detail);

    /// <summary>
    /// A day key as the portal writes it, from whatever a human typed.
    ///
    /// ⚠ "Monday", "MON", " mon " and "monday" all mean Monday. The simple editor writes `mon`; the
    /// advanced textarea is typed by a person, and refusing a whole week over a capital letter would be
    /// pedantry with a shop's opening times as the cost.
    /// </summary>
    private static string? DayKey(string raw)
    {
        var trimmed = raw.Trim().ToLowerInvariant();
        if (trimmed.Length < 3) return null;
        var k = trimmed[..3];
        return Days.Any(d => d.Key == k) ? k : null;
    }

    /// <summary>A day's value, which may be one span, several, a string, or an explicit "closed".</summary>
    private static (List<OpeningSpan> Spans, bool HadContent, bool Unreadable) SpansOf(JsonElement value)
    {
        var none = new List<OpeningSpan>();

        switch (value.ValueKind)
        {
            case JsonValueKind.Null or JsonValueKind.Undefined:
                return (none, false, false);

            // ⚠ `false` is how a person says shut. An explicit CLOSED, not junk.
            case JsonValueKind.False:
                return (none, true, false);
            case JsonValueKind.True:
                return (none, true, true);

            case JsonValueKind.String:
            {
                var text = value.GetString() ?? string.Empty;
                if (text.Trim().Length == 0
                    || text.Trim().Equals("closed", StringComparison.OrdinalIgnoreCase)
                    || text.Trim().Equals("shut", StringComparison.OrdinalIgnoreCase))
                    return (none, true, false);

                var span = SpanFromText(text);
                return span is null ? (none, true, false) : (new List<OpeningSpan> { span }, true, false);
            }

            case JsonValueKind.Array:
            {
                var items = value.EnumerateArray().ToList();
                // ⚠ `[]` IS "closed", deliberately — it is what the portal's editor leaves behind.
                if (items.Count == 0) return (none, true, false);

                var spans = items.Select(SpanFrom).Where(s => s is not null).Select(s => s!).ToList();
                return (spans, true, spans.Count == 0);
            }

            case JsonValueKind.Object:
            {
                var span = SpanFrom(value);
                return span is null ? (none, true, true) : (new List<OpeningSpan> { span }, true, false);
            }

            default:
                return (none, true, true);
        }
    }

    private static OpeningSpan? SpanFrom(JsonElement element)
    {
        if (element.ValueKind == JsonValueKind.String)
            return SpanFromText(element.GetString() ?? string.Empty);

        if (element.ValueKind != JsonValueKind.Object) return null;

        // ⚠ Case-insensitive field names for the same reason the keys are: this half of the blob can be
        // hand-typed. `from`/`to` are what somebody writes when they have not looked at the format.
        string? open = null, close = null;
        foreach (var field in element.EnumerateObject())
        {
            if (field.Value.ValueKind != JsonValueKind.String) continue;
            var name = field.Name.Trim().ToLowerInvariant();
            if (name is "open" or "from" or "start") open ??= Time(field.Value.GetString());
            else if (name is "close" or "to" or "end") close ??= Time(field.Value.GetString());
        }

        return open is null || close is null ? null : new OpeningSpan(open, close);
    }

    /// <summary>⚠ An en dash included: a range pasted out of a document carries one.</summary>
    private static OpeningSpan? SpanFromText(string text)
    {
        var parts = text.Split(new[] { '-', '–', '—' }, StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length != 2)
        {
            parts = text.Split(new[] { " to " }, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length != 2) return null;
        }

        var open = Time(parts[0]);
        var close = Time(parts[1]);
        return open is null || close is null ? null : new OpeningSpan(open, close);
    }

    /// <summary>
    /// A time, normalised for display.
    ///
    /// ⚠ `9:00` → `09:00` so the column lines up, and `09:00:00` → `09:00` because a shop's door does
    /// not open on a second boundary. Anything else is passed through untouched rather than rejected —
    /// a time this does not recognise is still better shown than swallowed.
    /// </summary>
    private static string? Time(string? raw)
    {
        var t = raw?.Trim();
        if (string.IsNullOrEmpty(t)) return null;

        var bits = t.Split(':');
        if (bits.Length is 2 or 3
            && int.TryParse(bits[0], out var h) && h is >= 0 and < 24
            && bits[1].Length == 2 && int.TryParse(bits[1], out var m) && m is >= 0 and < 60)
            return $"{h:00}:{m:00}";

        return t;
    }
}
